### PR TITLE
forkusernsexec: close file descriptors before exec

### PR DESCRIPTION
**Pull request by brauner**
Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

File: lxd/main_forkusernsexec.go [hunk 1]
```diff
@@ -221,6 +222,10 @@ out_reap:
 	if (!argvp || !*argvp)
 		return log_error(EXIT_FAILURE, "No command specified");
 
+	ret = lxd_close_range(FD_PIPE_GIDMAP + 1, UINT_MAX, CLOSE_RANGE_CLOEXEC);
```
Comment (tomponline) [inline lxd/main_forkusernsexec.go:225]: Should this be FD_PIPE_UIDMAP+1?

Also, I am going to need to pass additional FDs into this command, in addition to 3 and 4, so just want to check the additional ones won't get closed?

Comment (brauner) [reply lxd/main_forkusernsexec.go:225]: > Should this be FD_PIPE_UIDMAP+1?

It needs to be FD_PIPE_GIDMAP and +1 as the fd passed as the first argument will be closed as well.

> 
> Also, I am going to need to pass additional FDs into this command, in addition to 3 and 4, so just want to check the additional ones won't get closed?

You need the program you're execing to inherit fds? If that's the case we might just not do this check at all.

Comment (tomponline) [reply lxd/main_forkusernsexec.go:225]: Yes I will need to pass, at least, 1 other file handle in additional to FD_PIPE_UIDMAP and FD_PIPE_GIDMAP

Comment (tomponline) [reply lxd/main_forkusernsexec.go:225]: See https://github.com/lxc/lxd/blob/aca7736aca79b882fdd43bd02eaf4053b2a17676/lxd/device/device_utils_disk.go#L298 where I will wrap that call with the forkusernsexec command, and then pass fd 5 as the socket, changing the `--fd 3` arg to `--fd 5`.

Comment (tomponline) [reply lxd/main_forkusernsexec.go:225]: It works right now btw :)

**Comment by stgraber** [comment]
@brauner failing to build

**Comment by tomponline** [comment]
Do we need this anyway?

**Comment by stgraber** [comment]
Well, we should close the two range FDs at least (3 and 4), the rest we probably shouldn't touch though.

**Review by tomponline — APPROVED** [review]

**Comment by brauner** [comment]
Well definitely need something like this. We will not risk leaking fds from the LXD daemon by accident even if it is a far-fetched attack. This can be super bad. I think we'll just let the caller tell us via an env variable what the largest open fd is they expect.

File: lxd/main_forkusernsexec.go [hunk 2]
```diff
@@ -233,14 +273,16 @@ void forkusernsexec(void)
...
+	global     *cmdGlobal
+	keepFdUpTo int
 }
 
 func (c *cmdForkusernsexec) Command() *cobra.Command {
 	// Main subcommand
 	cmd := &cobra.Command{}
 	cmd.Use = "forkusernsexec <cmd> <args>"
 	cmd.Short = "Run command in user namespace"
+	cmd.Flags().IntVar(&c.keepFdUpTo, "keep-fd-up-to", (1 << 31), fmt.Sprintf("Keep all fds below this one and close all the ones above it"))
```
Comment (tomponline) [inline lxd/main_forkusernsexec.go:285]: This comment doesn't define what happens to the number specified in the argument, only what happens to the fds below it and above it. I think tweak it to say something like:

"Keep all fds below and including this one and close all the ones above it"

Comment (brauner) [reply lxd/main_forkusernsexec.go]: done

**Review by tomponline — APPROVED** [review]

**Pull request merged**